modelservice/src/main/java/com/expedia/adaptivealerting/modelservice/dao/es/ElasticSearchDetectorMappingService.java

```diff
@@ -0,0 +1,377 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import static com.expedia.adaptivealerting.modelservice.dao.es.DetectorMappingEntity.LAST_MOD_TIME_KEYWORD;
+
+@Component
+@Slf4j
+@SuppressWarnings({"PMD.ExcessiveImports", "PMD.AvoidThrowingRawExceptionTypes"})
+public class ElasticSearchDetectorMappingService implements DetectorMappingService {
+    private static ObjectMapper objectmapper    = new ObjectMapper();
+
+    @Autowired
+    private ElasticSearchConfig elasticSearchConfig;
+
+    @Autowired
+    private ElasticSearchClient elasticSearchClient;
+
+    private final Timer delayTimer;
+    private final Counter exceptionCount;
+
+    @Autowired
+    public ElasticSearchDetectorMappingService(MetricRegistry metricRegistry) {
+        this.delayTimer = metricRegistry.timer("es-lookup.time-delay");
+        this.exceptionCount = metricRegistry.counter("es-lookup.exception");
+    }
+
+    @Override
+    public MatchingDetectorsResponse findMatchingDetectorMappings(List<Map<String, String>> tagsList) {
+        try {
+            List<BytesReference> refList = new ArrayList<>();
+            for (Map<String, String> tags: tagsList) {
+                XContentBuilder xContent = XContentFactory.jsonBuilder();
+                xContent.map(tags);
+                refList.add(BytesReference.bytes(xContent));
+            }
+            PercolateQueryBuilder percolateQuery =
+                new PercolateQueryBuilder(DetectorMappingEntity.QUERY_KEYWORD, elasticSearchConfig.getDocType(),
+                    refList, XContentType.JSON);
+            SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+            ConstantScoreQueryBuilder constantScoreQueryBuilder = new ConstantScoreQueryBuilder(percolateQuery);
+            searchSourceBuilder.query(constantScoreQueryBuilder);
+            searchSourceBuilder.timeout(new TimeValue(elasticSearchConfig.getConnectionTimeout()));
+            //FIXME setting default result set size to 500.
+            searchSourceBuilder.size(500);
+            return getDetectorMappings(searchSourceBuilder, tagsList);
+        } catch (IOException e) {
+            this.exceptionCount.inc();
+            log.error("Error ES lookup", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public String createDetectorMapping(CreateDetectorMappingRequest createRequest) {
+        Set<String> fields = getFields(createRequest);
+        Set<String> newFieldMappings = getFieldsWithoutExistingMapping(fields);
+        updateIndexMappings(newFieldMappings);
+        final DetectorMappingEntity detectorMappingEntity = new DetectorMappingEntity(createRequest.getUser(),
+            createRequest.getDetector(), QueryUtil.buildQuery(createRequest.getExpression()), true,
+            System.currentTimeMillis(), System.currentTimeMillis());
```

Comment:
Less than ideal to create new instances using a large number of parameters. Consider replacing the `@AllArgsConstructor` on the `DetectorMappingEntity` with Lombok's `@Accessors(chain = true)`.

Reply:
Updated as advised.